File: frontend/src/api/messages.ts
import http from './index'

export interface MessageLog {
  id: string
  session_id: string
  direction: 'request' | 'response' | 'notification'
  timestamp: string
  method?: string
  request_id?: number | string
  data: any
  duration_ms?: number
  error?: any
}

export interface MessageStats {
  total: number
  requests: number
  responses: number
  notifications: number
  errors: number
  methods: Record<string, number>
  avg_duration_ms: number
  max_duration_ms: number
  min_duration_ms: number
}

export const messagesApi = {
  query: (params?: { direction?: string; method?: string; search_text?: string; limit?: number }) =>
    http.get<any, { messages: MessageLog[]; total: number }>('/api/messages', { params }),
  stats: () => http.get<any, MessageStats>('/api/messages/stats'),
  clear: () => http.delete('/api/messages'),
  exportJson: () => http.get('/api/messages/export/json', { responseType: 'blob' }),
  exportCsv: () => http.get('/api/messages/export/csv', { responseType: 'blob' }),
  history: () => http.get<any, { files: Array<{ name: string; size: number; modified: string }> }>('/api/messages/history'),
  getHistory: (filename: string) => http.get<any, MessageLog[]>(`/api/messages/history/${filename}`),
}

File: backend/api/messages.py
from datetime import datetime
from pathlib import Path

from fastapi import APIRouter, HTTPException
from fastapi.responses import Response

from ..models.message_log import MessageFilter
from .state import app_state

router = APIRouter()


@router.get("")
async def query_messages(
    direction: str = None,
    method: str = None,
    search_text: str = None,
    limit: int = 500,
):
    f = MessageFilter(
        direction=direction,
        method=method,
        search_text=search_text,
        limit=limit,
    )
    messages = app_state.message_logger.query(f)
    return {
        "messages": [m.model_dump(mode="json") for m in messages],
        "total": app_state.message_logger.count,
    }


@router.get("/stats")
async def message_stats():
    return app_state.message_logger.get_stats()


@router.delete("")
async def clear_messages():
    app_state.message_logger.clear()
    return {"cleared": True}


@router.get("/export/json")
async def export_json():
    data = app_state.message_logger.export_json()
    return Response(content=data, media_type="application/json")


@router.get("/export/csv")
async def export_csv():
    data = app_state.message_logger.export_csv()
    return Response(content=data, media_type="text/csv")


@router.get("/history")
async def list_history():
    history_dir = Path("data/history")
    if not history_dir.exists():
        return {"files": []}
    files = sorted(history_dir.glob("messages_*.json"), reverse=True)
    return {
        "files": [
            {
                "name": f.name,
                "size": f.stat().st_size,
                "modified": datetime.fromtimestamp(f.stat().st_mtime).isoformat(),
            }
            for f in files[:50]
        ]
    }


@router.get("/history/{filename}")
async def get_history(filename: str):
    if ".." in filename or "/" in filename:
        raise HTTPException(400, "Invalid filename")
    filepath = Path("data/history") / filename
    if not filepath.exists():
        raise HTTPException(404, "History file not found")
    data = filepath.read_text(encoding="utf-8")
    return Response(content=data, media_type="application/json")
